File: python-adk-runtime/src/detectviz_adk/config/loader.py
# Configuration loader for Python ADK Runtime
# 對齊 Detectviz SSOT：統一搜尋順序、環境變數覆蓋、Schema 驗證
# 生效優先序（高→低）：
#  1. 函式參數 path / CLI --config
#  2. 環境變數 DETECTVIZ_CONFIG_FILE
#  3. `./config.yaml`（工作目錄）
#  4. `../contracts/config.yaml`（團隊覆蓋，可選）
#  5. `./contracts/samples/config.yaml`（SSOT 樣本兜底）

from __future__ import annotations

import os
import json
import yaml
import jsonschema
import logging
from typing import Dict, Any, Optional, List, Tuple
from pathlib import Path

logger = logging.getLogger(__name__)

# ---- Public API -----------------------------------------------------------

def load_config(path: Optional[str] = None) -> Dict[str, Any]:
    """載入 Detectviz 統一設定並通過 Schema 驗證。

    1) 解析設定檔（依優先序尋找）
    2) 套用預設值（僅補缺，不覆蓋）
    3) 環境變數覆蓋（DETECTVIZ__* 鍵位對齊 Go）
    4.以 contracts/schemas/config.schema.json 驗證
    """
    resolved, tried = _resolve_config_path(path)

    config: Dict[str, Any] = {}
    if resolved:
        with open(resolved, "r", encoding="utf-8") as f:
            config = yaml.safe_load(f) or {}
        logger.info("Loaded configuration", extra={"config.path": str(resolved)})
    else:
        logger.warning("No config file found; using defaults only", extra={"tried": tried})

    config = _apply_defaults(config)
    _apply_env_overrides(config)

    _validate_config(config)

    # 脫敏摘要輸出
    logger.info(
        "Configuration effective",
        extra={
            "env": config.get("env"),
            "observability.mode": _dget(config, "observability.mode"),
            "otlp.protocol": _dget(config, "observability.otlp.protocol"),
            "otlp.endpoint": _dget(config, "observability.otlp.endpoint"),
        },
    )

    return config


def get_toolbridge_addr(default: str = "127.0.0.1:5002") -> str:
    """取得 Python RemoteTool 連線位址（不寫入 SSOT Config，以環境變數為主）。"""
    return os.getenv("DETECTVIZ_TOOLBRIDGE_ADDR", default)


# ---- Internal helpers -----------------------------------------------------

_DEF: Dict[str, Any] = {
    "env": "development",
    "observability": {
        "mode": "lgtm_local",
        "otlp": {
            "protocol": "grpc",
            "endpoint": "127.0.0.1:4317",
            "insecure": True,
            "headers": {},
        },
        # Python 僅傳遞 span context；logs/pprof 由 Go + Alloy 負責
        "logs": {"mode": "off", "file": {"path": ""}},
        "profiling": {"enabled": False, "pprof_address": "", "application_name": "", "tags": {}},
        "resource": {
            "service_name": "detectviz-adk-runtime",
            "service_version": "0.1.0",
            "environment": "dev",
        },
        "sampling": {"ratio": 1.0},
    },
    "plugin": {"paths": [], "registry": ""},
    "memory": {"backend": "inmem", "dsn": "", "default_ttl_seconds": 3600},
}


def _apply_defaults(cfg: Dict[str, Any]) -> Dict[str, Any]:
    return _deep_merge(_DEF, cfg)


def _apply_env_overrides(cfg: Dict[str, Any]) -> None:
    """以 DETECTVIZ_* 規範覆蓋設定（就地修改）。

    對齊 Go 端鍵位：
      - DETECTVIZ_ENV → env
      - DETECTVIZ__OBSERVABILITY__MODE → observability.mode
      - DETECTVIZ__OBSERVABILITY__OTLP__PROTOCOL → observability.otlp.protocol
      - DETECTVIZ__OBSERVABILITY__OTLP__ENDPOINT → observability.otlp.endpoint
      - DETECTVIZ__OBSERVABILITY__OTLP__INSECURE → observability.otlp.insecure
      - DETECTVIZ__OBSERVABILITY__OTLP__HEADERS → observability.otlp.headers (k=v,k2=v2)
      - DETECTVIZ__OBSERVABILITY__RESOURCE__SERVICE_NAME → observability.resource.service_name
      - DETECTVIZ__OBSERVABILITY__RESOURCE__SERVICE_VERSION → observability.resource.service_version
      - DETECTVIZ__OBSERVABILITY__RESOURCE__ENVIRONMENT → observability.resource.environment
      - DETECTVIZ__OBSERVABILITY__SAMPLING__RATIO → observability.sampling.ratio
      - DETECTVIZ__PLUGIN__PATHS → plugin.paths (csv)
      - DETECTVIZ__PLUGIN__REGISTRY → plugin.registry
      - DETECTVIZ__MEMORY__BACKEND → memory.backend
      - DETECTVIZ__MEMORY__DSN → memory.dsn
      - DETECTVIZ__MEMORY__DEFAULT_TTL_SECONDS → memory.default_ttl_seconds
    """

    mapping = {
        "DETECTVIZ_ENV": ("env", _as_str),
        # Observability / OTLP
        "DETECTVIZ__OBSERVABILITY__MODE": ("observability.mode", _as_str),
        "DETECTVIZ__OBSERVABILITY__OTLP__PROTOCOL": ("observability.otlp.protocol", _as_str),
        "DETECTVIZ__OBSERVABILITY__OTLP__ENDPOINT": ("observability.otlp.endpoint", _as_str),
        "DETECTVIZ__OBSERVABILITY__OTLP__INSECURE": ("observability.otlp.insecure", _as_bool),
        "DETECTVIZ__OBSERVABILITY__OTLP__HEADERS": ("observability.otlp.headers", _as_headers),
        # Resource & Sampling
        "DETECTVIZ__OBSERVABILITY__RESOURCE__SERVICE_NAME": ("observability.resource.service_name", _as_str),
        "DETECTVIZ__OBSERVABILITY__RESOURCE__SERVICE_VERSION": ("observability.resource.service_version", _as_str),
        "DETECTVIZ__OBSERVABILITY__RESOURCE__ENVIRONMENT": ("observability.resource.environment", _as_str),
        "DETECTVIZ__OBSERVABILITY__SAMPLING__RATIO": ("observability.sampling.ratio", _as_float),
        # Plugin & Memory
        "DETECTVIZ__PLUGIN__PATHS": ("plugin.paths", _as_csv),
        "DETECTVIZ__PLUGIN__REGISTRY": ("plugin.registry", _as_str),
        "DETECTVIZ__MEMORY__BACKEND": ("memory.backend", _as_str),
        "DETECTVIZ__MEMORY__DSN": ("memory.dsn", _as_str),
        "DETECTVIZ__MEMORY__DEFAULT_TTL_SECONDS": ("memory.default_ttl_seconds", _as_int),
    }

    for env, (keypath, caster) in mapping.items():
        val = os.getenv(env)
        if val is None or val == "":
            continue
        try:
            casted = caster(val)
            _dset(cfg, keypath, casted)
            logger.info("Env override", extra={"env": env, "key": keypath})
        except Exception as e:
            logger.warning("Failed to apply env override", extra={"env": env, "err": str(e)})


def _validate_config(cfg: Dict[str, Any]) -> None:
    schema_path = _resolve_schema_path()
    if not schema_path:
        logger.warning("Configuration schema not found, skip validation")
        return

    with open(schema_path, "r", encoding="utf-8") as f:
        schema = json.load(f)

    try:
        jsonschema.Draft202012Validator(schema).validate(cfg)
        logger.info("Configuration validated", extra={"schema": str(schema_path)})
    except jsonschema.ValidationError as e:
        raise ValueError(f"Invalid configuration: {e.message}") from e


# ---- Path resolvers -------------------------------------------------------

def _resolve_config_path(arg: Optional[str]) -> Tuple[Optional[Path], List[str]]:
    tried: List[str] = []

    # 1) explicit arg
    if arg and str(arg).strip():
        p = Path(arg)
        tried.append(str(p))
        if p.exists() and p.is_file():
            return p, tried
        return None, tried

    # 2) env
    env_path = os.getenv("DETECTVIZ_CONFIG_FILE")
    if env_path:
        p = Path(env_path)
        tried.append(str(p))
        if p.exists() and p.is_file():
            return p, tried

    # 3) ./config.yaml
    for cand in (Path("./config.yaml"), Path("config.yaml")):
        tried.append(str(cand))
        if cand.exists() and cand.is_file():
            return cand, tried

    # 4../contracts/config.yaml
    for cand in (Path("./contracts/config.yaml"), Path("contracts/config.yaml")):
        tried.append(str(cand))
        if cand.exists() and cand.is_file():
            return cand, tried

    # 5) ./contracts/samples/config.yaml
    for cand in (
        Path("./contracts/samples/config.yaml"),
        Path("contracts/samples/config.yaml"),
    ):
        tried.append(str(cand))
        if cand.exists() and cand.is_file():
            return cand, tried

    return None, tried


def _resolve_schema_path() -> Optional[Path]:
    # 掃描常見位置（以 repo 根為準）
    candidates = [
        Path("./contracts/schemas/config.schema.json"),
        Path("contracts/schemas/config.schema.json"),
        Path(__file__).resolve().parents[4] / "contracts/schemas/config.schema.json",
        Path(__file__).resolve().parents[3] / "contracts/schemas/config.schema.json",
    ]
    for p in candidates:
        if p.exists():
            return p
    return None


# ---- Generic utils --------------------------------------------------------

def _deep_merge(base: Dict[str, Any], override: Dict[str, Any]) -> Dict[str, Any]:
    out = dict(base)
    for k, v in (override or {}).items():
        if isinstance(v, dict) and isinstance(out.get(k), dict):
            out[k] = _deep_merge(out[k], v)
        else:
            out[k] = v
    return out


def _dget(d: Dict[str, Any], dotted: str, default: Any = None) -> Any:
    cur = d
    for part in dotted.split("."):
        if not isinstance(cur, dict) or part not in cur:
            return default
        cur = cur[part]
    return cur


def _dset(d: Dict[str, Any], dotted: str, value: Any) -> None:
    cur = d
    parts = dotted.split(".")
    for p in parts[:-1]:
        if p not in cur or not isinstance(cur[p], dict):
            cur[p] = {}
        cur = cur[p]
    cur[parts[-1]] = value


# casters

def _as_str(v: str) -> str:
    return str(v)


def _as_int(v: str) -> int:
    return int(v.strip())


def _as_float(v: str) -> float:
    return float(v.strip())


def _as_bool(v: str) -> bool:
    s = v.strip().lower()
    if s in ("1", "true", "t", "yes", "y", "on"):
        return True
    if s in ("0", "false", "f", "no", "n", "off"):
        return False
    raise ValueError(f"invalid bool: {v}")


def _as_csv(v: str) -> list:
    return [p.strip() for p in v.split(",") if p.strip()]


def _as_headers(v: str) -> Dict[str, str]:
    # 例如："authorization=Bearer xxx, x-foo=bar"
    out: Dict[str, str] = {}
    for part in v.split(","):
        part = part.strip()
        if not part:
            continue
        if "=" not in part:
            continue
        k, val = part.split("=", 1)
        k = k.strip()
        val = val.strip()
        if k:
            out[k] = val
    return out


# ---- Emergency minimal ----------------------------------------------------

def _get_emergency_config() -> Dict[str, Any]:
    return {
        "env": "emergency",
        "observability": {
            "mode": "lgtm_local",
            "otlp": {"protocol": "grpc", "endpoint": "127.0.0.1:4317", "insecure": True},
            "resource": {"service_name": "detectviz-adk-runtime", "service_version": "0.0.0", "environment": "emerg"},
        },
        "plugin": {"paths": []},
        "memory": {"backend": "inmem"},
    }

File: go-platform/internal/configx/loader.go
// Package configx provides configuration loading and validation against contracts schemas
// This adheres to spec.md requirement: "Go 端負責 Config 驗證與載入"
package configx

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/xeipuuv/gojsonschema"
	"go.uber.org/zap"
	"gopkg.in/yaml.v3"
)

// Config represents the platform configuration structure (SSOT-aligned, no backward-compat fields)
// observability 對齊 otel_init.go 的讀取模式：
//   observability:
//     mode: lgtm_local|grafana_cloud|gcp
//     otlp:
//       protocol: grpc|http
//       endpoint: host:port (grpc) | http(s)://host:port (http)
//       insecure: true|false
//       headers: { key: value }
//     logs:
//       mode: file|stdout|off
//       file:
//         path: ./var/log/detectviz/detectviz.log
//         max_size_mb: 50
//         max_backups: 7
//         max_age_days: 14
//         compress: true
//     profiling:
//       enabled: true|false
//       pprof_address: 127.0.0.1:6060
//       application_name: go-platform
//       tags: { service.name: go-platform, deployment.environment: dev }
//   resource/sampling 為可選，用於後續擴充
// 其他區塊維持最小可運行：grpc / plugin / memory

type Config struct {
	Env string `yaml:"env" json:"env"`

	Observability struct {
		Mode string `yaml:"mode" json:"mode"`

		OTLP struct {
			Protocol string            `yaml:"protocol" json:"protocol"` // grpc|http
			Endpoint string            `yaml:"endpoint" json:"endpoint"`
			Insecure bool              `yaml:"insecure" json:"insecure"`
			Headers  map[string]string `yaml:"headers" json:"headers"`
		} `yaml:"otlp" json:"otlp"`

		Logs struct {
			Mode string `yaml:"mode" json:"mode"` // file|stdout|off
			File struct {
				Path       string `yaml:"path" json:"path"`
				MaxSizeMB  int    `yaml:"max_size_mb" json:"max_size_mb"`
				MaxBackups int    `yaml:"max_backups" json:"max_backups"`
				MaxAgeDays int    `yaml:"max_age_days" json:"max_age_days"`
				Compress   bool   `yaml:"compress" json:"compress"`
			} `yaml:"file" json:"file"`
		} `yaml:"logs" json:"logs"`

		Resource struct {
			ServiceName    string `yaml:"service_name" json:"service_name"`
			ServiceVersion string `yaml:"service_version" json:"service_version"`
			Environment    string `yaml:"environment" json:"environment"`
		} `yaml:"resource" json:"resource"`

		Sampling struct {
			Ratio float64 `yaml:"ratio" json:"ratio"`
		} `yaml:"sampling" json:"sampling"`

		Profiling struct {
			Enabled         bool              `yaml:"enabled" json:"enabled"`
			PprofAddress    string            `yaml:"pprof_address" json:"pprof_address"` // 127.0.0.1:6060
			ApplicationName string            `yaml:"application_name" json:"application_name"`
			Tags            map[string]string `yaml:"tags" json:"tags"`
		} `yaml:"profiling" json:"profiling"`
	} `yaml:"observability" json:"observability"`

	GRPC struct {
		Listen       string `yaml:"listen" json:"listen"`
		MaxRecvBytes int    `yaml:"max_recv_bytes" json:"max_recv_bytes"`
		MaxSendBytes int    `yaml:"max_send_bytes" json:"max_send_bytes"`
		TLS          struct {
			Enabled  bool   `yaml:"enabled" json:"enabled"`
			CertFile string `yaml:"cert_file" json:"cert_file"`
			KeyFile  string `yaml:"key_file" json:"key_file"`
		} `yaml:"tls" json:"tls"`
	} `yaml:"grpc" json:"grpc"`

	Plugin struct {
		Paths    []string `yaml:"paths" json:"paths"`
		Registry string   `yaml:"registry" json:"registry"`
	} `yaml:"plugin" json:"plugin"`

	Memory struct {
		Backend           string `yaml:"backend" json:"backend"` // inmem|redis|weaviate|chroma|vertex
		DSN               string `yaml:"dsn" json:"dsn"`
		DefaultTTLSeconds int    `yaml:"default_ttl_seconds" json:"default_ttl_seconds"`
	} `yaml:"memory" json:"memory"`
}

// LoadAndValidate loads configuration file with unified precedence and validates against SSOT schema.
// Precedence (highest → lowest):
//  1. explicit argument --config
//  2. env DETECTVIZ_CONFIG_FILE
//  3. ./config.yaml (working directory)
//  4. ./contracts/config.yaml (team-shared overlay)
//  5. ./contracts/samples/config.yaml (SSOT sample; last resort)
func LoadAndValidate(configPath string) (*Config, error) {
	resolved, tried, err := resolveConfigPath(configPath)
	if err != nil {
		return nil, err
	}

	// Read configuration file
	data, err := os.ReadFile(resolved)
	if err != nil {
		return nil, fmt.Errorf("failed to read config file %s: %w", resolved, err)
	}

	var cfg Config
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return nil, fmt.Errorf("failed to parse YAML config (%s): %w", resolved, err)
	}

	applyDefaults(&cfg)

	// 環境變數覆蓋（與 Python 對齊）
	applyEnvOverrides(&cfg)

	// Validate against contracts schema
	if err := validateWithSchema(&cfg); err != nil {
		return nil, fmt.Errorf("config validation failed: %w", err)
	}

	zap.L().Info("Configuration loaded and validated",
		zap.String("config.path", resolved),
		zap.Strings("config.search_tried", tried),
		zap.String("env", cfg.Env),
		zap.String("observability.mode", cfg.Observability.Mode),
		zap.String("otlp.protocol", cfg.Observability.OTLP.Protocol),
		zap.String("otlp.endpoint", cfg.Observability.OTLP.Endpoint),
	)

	return &cfg, nil
}

// resolveConfigPath resolves the effective config path according to the unified precedence.
// It returns (resolvedPath, triedCandidates, error).
func resolveConfigPath(arg string) (string, []string, error) {
	// If user provided an explicit path via flag, honor it strictly.
	if strings.TrimSpace(arg) != "" {
		abs := arg
		if !filepath.IsAbs(abs) {
			if a, err := filepath.Abs(abs); err == nil {
				abs = a
			}
		}
		if st, err := os.Stat(abs); err == nil && !st.IsDir() {
			return abs, []string{abs}, nil
		}
		return "", []string{abs}, fmt.Errorf("config file not found: %s", abs)
	}

	tried := []string{}

	// env override
	if env := strings.TrimSpace(os.Getenv("DETECTVIZ_CONFIG_FILE")); env != "" {
		tried = append(tried, env)
		if st, err := os.Stat(env); err == nil && !st.IsDir() {
			return env, tried, nil
		}
	}

	// working dir
	candidates := []string{"./config.yaml", "config.yaml"}
	for _, p := range candidates {
		tried = append(tried, p)
		if st, err := os.Stat(p); err == nil && !st.IsDir() {
			return p, tried, nil
		}
	}

	// contracts-based fallbacks
	if cd := findContractsDir(); cd != "" {
		p1 := filepath.Join(cd, "config.yaml")
		p2 := filepath.Join(cd, "samples", "config.yaml")
		for _, p := range []string{p1, p2} {
			tried = append(tried, p)
			if st, err := os.Stat(p); err == nil && !st.IsDir() {
				return p, tried, nil
			}
		}
	}

	return "", tried, fmt.Errorf("no config file found; tried %v", tried)
}

// validateWithSchema validates config against contracts/schemas/config.schema.json
func validateWithSchema(cfg *Config) error {
	// Find contracts directory - should be a sibling of go-platform
	contractsDir := findContractsDir()
	if contractsDir == "" {
		return fmt.Errorf("contracts directory not found")
	}

	schemaPath := filepath.Join(contractsDir, "schemas", "config.schema.json")

	// Check if schema file exists
	if _, err := os.Stat(schemaPath); os.IsNotExist(err) {
		return fmt.Errorf("config schema not found at %s", schemaPath)
	}

	schemaLoader := gojsonschema.NewReferenceLoader("file://" + schemaPath)
	documentLoader := gojsonschema.NewGoLoader(cfg)

	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
	if err != nil {
		return fmt.Errorf("schema validation error: %w", err)
	}

	if !result.Valid() {
		var errors []string
		for _, desc := range result.Errors() {
			errors = append(errors, fmt.Sprintf("  - %s", desc.String()))
		}
		return fmt.Errorf("configuration validation failed:\n%v", errors)
	}

	return nil
}

// findContractsDir locates the contracts directory relative to go-platform
func findContractsDir() string {
	// Try relative path from go-platform
	paths := []string{
		"contracts",
		"../contracts",       // From go-platform root
		"../../contracts",    // From go-platform/internal
		"../../../contracts", // From go-platform/internal/configx
	}

	for _, path := range paths {
		absPath, err := filepath.Abs(path)
		if err != nil {
			continue
		}
		if _, err := os.Stat(absPath); err == nil {
			return absPath
		}
	}

	return ""
}

// applyDefaults sets default values according to spec.md requirements
func applyDefaults(cfg *Config) {
	// Ensure non-nil maps/slices for schema validation
	if cfg.Observability.OTLP.Headers == nil {
		cfg.Observability.OTLP.Headers = map[string]string{}
	}
	if cfg.Plugin.Paths == nil {
		cfg.Plugin.Paths = []string{}
	}
	if cfg.Plugin.Registry == "" {
		cfg.Plugin.Registry = "file"
	}
	// gRPC payload limits must be >= 1MB (schema minimum)
	if cfg.GRPC.MaxRecvBytes < 1048576 {
		cfg.GRPC.MaxRecvBytes = 4 * 1024 * 1024
	}
	if cfg.GRPC.MaxSendBytes < 1048576 {
		cfg.GRPC.MaxSendBytes = 4 * 1024 * 1024
	}

	// Default observability mode
	if cfg.Observability.Mode == "" {
		cfg.Observability.Mode = "lgtm_local"
	}

	// Default OTLP protocol & endpoint
	if cfg.Observability.OTLP.Protocol == "" {
		cfg.Observability.OTLP.Protocol = "grpc"
	}
	if cfg.Observability.OTLP.Endpoint == "" {
		if cfg.Observability.OTLP.Protocol == "http" {
			cfg.Observability.OTLP.Endpoint = "http://localhost:4318"
		} else {
			cfg.Observability.OTLP.Endpoint = "localhost:4317"
		}
	}
	// Default logs
	if cfg.Observability.Logs.Mode == "" {
		cfg.Observability.Logs.Mode = "file"
	}
	if cfg.Observability.Logs.File.Path == "" {
		cfg.Observability.Logs.File.Path = "./var/log/detectviz/detectviz.log"
	}
	if cfg.Observability.Logs.File.MaxSizeMB == 0 {
		cfg.Observability.Logs.File.MaxSizeMB = 50
	}
	if cfg.Observability.Logs.File.MaxBackups == 0 {
		cfg.Observability.Logs.File.MaxBackups = 7
	}
	if cfg.Observability.Logs.File.MaxAgeDays == 0 {
		cfg.Observability.Logs.File.MaxAgeDays = 14
	}
	// compress 預設 true（YAML 未設定時為零值 false，這裡置 true）
	if !cfg.Observability.Logs.File.Compress {
		cfg.Observability.Logs.File.Compress = true
	}

	// Default gRPC settings（ToolBridge）
	if cfg.GRPC.Listen == "" {
		cfg.GRPC.Listen = "0.0.0.0:6606"
	}

	// Default memory backend
	if cfg.Memory.Backend == "" {
		cfg.Memory.Backend = "inmem"
	}

	// Default profiling（僅支援 pprof）
	if cfg.Observability.Profiling.PprofAddress == "" {
		cfg.Observability.Profiling.PprofAddress = "127.0.0.1:6060"
	}
	if cfg.Observability.Profiling.ApplicationName == "" {
		cfg.Observability.Profiling.ApplicationName = "go-platform"
	}
	if cfg.Observability.Profiling.Tags == nil {
		cfg.Observability.Profiling.Tags = map[string]string{}
	}
}

// GetObservabilityConfig returns a nested map (for otel_init.go)
func (c *Config) GetObservabilityConfig() map[string]interface{} {
	config := map[string]interface{}{
		"mode": c.Observability.Mode,
		"otlp": map[string]interface{}{
			"protocol": c.Observability.OTLP.Protocol,
			"endpoint": c.Observability.OTLP.Endpoint,
			"insecure": c.Observability.OTLP.Insecure,
			"headers":  c.Observability.OTLP.Headers,
		},
		"logs": map[string]interface{}{
			"mode": c.Observability.Logs.Mode,
			"file": map[string]interface{}{
				"path":         c.Observability.Logs.File.Path,
				"max_size_mb":  c.Observability.Logs.File.MaxSizeMB,
				"max_backups":  c.Observability.Logs.File.MaxBackups,
				"max_age_days": c.Observability.Logs.File.MaxAgeDays,
				"compress":     c.Observability.Logs.File.Compress,
			},
		},
		"resource": map[string]string{
			"service.name":           c.Observability.Resource.ServiceName,
			"service.version":        c.Observability.Resource.ServiceVersion,
			"deployment.environment": c.Observability.Resource.Environment,
		},
		"sampling": c.Observability.Sampling.Ratio,
	}

	// Add profiling configuration if enabled（僅 pprof）
	if c.Observability.Profiling.Enabled {
		config["profiling"] = map[string]interface{}{
			"enabled":          c.Observability.Profiling.Enabled,
			"pprof_address":    c.Observability.Profiling.PprofAddress,
			"application_name": c.Observability.Profiling.ApplicationName,
			"tags":             c.Observability.Profiling.Tags,
		}
	}

	return config
}

// applyEnvOverrides 以 DETECTVIZ_ 前綴環境變數覆蓋設定
func applyEnvOverrides(c *Config) {
	setStr(&c.Env, os.Getenv("DETECTVIZ_ENV"))

	// gRPC
	setStr(&c.GRPC.Listen, os.Getenv("DETECTVIZ__GRPC__LISTEN"))
	if v := os.Getenv("DETECTVIZ__GRPC__MAX_RECV_BYTES"); v != "" {
		if iv, err := atoi(v); err == nil {
			c.GRPC.MaxRecvBytes = iv
		}
	}
	if v := os.Getenv("DETECTVIZ__GRPC__MAX_SEND_BYTES"); v != "" {
		if iv, err := atoi(v); err == nil {
			c.GRPC.MaxSendBytes = iv
		}
	}

	// Observability base
	setStr(&c.Observability.Mode, os.Getenv("DETECTVIZ__OBSERVABILITY__MODE"))
	setStr(&c.Observability.OTLP.Protocol, os.Getenv("DETECTVIZ__OBSERVABILITY__OTLP__PROTOCOL"))
	setStr(&c.Observability.OTLP.Endpoint, os.Getenv("DETECTVIZ__OBSERVABILITY__OTLP__ENDPOINT"))
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__OTLP__INSECURE"); v != "" {
		if bv, err := parseBool(v); err == nil {
			c.Observability.OTLP.Insecure = bv
		}
	}
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__OTLP__HEADERS"); v != "" {
		c.Observability.OTLP.Headers = parseHeadersCSV(v)
	}

	// Logs
	setStr(&c.Observability.Logs.Mode, os.Getenv("DETECTVIZ__OBSERVABILITY__LOGS__MODE"))
	setStr(&c.Observability.Logs.File.Path, os.Getenv("DETECTVIZ__OBSERVABILITY__LOGS__FILE__PATH"))
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__LOGS__FILE__MAX_SIZE_MB"); v != "" {
		if iv, err := atoi(v); err == nil {
			c.Observability.Logs.File.MaxSizeMB = iv
		}
	}
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__LOGS__FILE__MAX_BACKUPS"); v != "" {
		if iv, err := atoi(v); err == nil {
			c.Observability.Logs.File.MaxBackups = iv
		}
	}
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__LOGS__FILE__MAX_AGE_DAYS"); v != "" {
		if iv, err := atoi(v); err == nil {
			c.Observability.Logs.File.MaxAgeDays = iv
		}
	}
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__LOGS__FILE__COMPRESS"); v != "" {
		if bv, err := parseBool(v); err == nil {
			c.Observability.Logs.File.Compress = bv
		}
	}

	// Profiling (pprof)
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__PROFILING__ENABLED"); v != "" {
		if bv, err := parseBool(v); err == nil {
			c.Observability.Profiling.Enabled = bv
		}
	}
	setStr(&c.Observability.Profiling.PprofAddress, os.Getenv("DETECTVIZ__OBSERVABILITY__PROFILING__PPROF_ADDRESS"))
	setStr(&c.Observability.Profiling.ApplicationName, os.Getenv("DETECTVIZ__OBSERVABILITY__PROFILING__APPLICATION_NAME"))
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__PROFILING__TAGS"); v != "" {
		c.Observability.Profiling.Tags = parseHeadersCSV(v)
	}

	// Resource
	setStr(&c.Observability.Resource.ServiceName, os.Getenv("DETECTVIZ__OBSERVABILITY__RESOURCE__SERVICE_NAME"))
	setStr(&c.Observability.Resource.ServiceVersion, os.Getenv("DETECTVIZ__OBSERVABILITY__RESOURCE__SERVICE_VERSION"))
	setStr(&c.Observability.Resource.Environment, os.Getenv("DETECTVIZ__OBSERVABILITY__RESOURCE__ENVIRONMENT"))

	// Sampling
	if v := os.Getenv("DETECTVIZ__OBSERVABILITY__SAMPLING__RATIO"); v != "" {
		if fv, err := atof(v); err == nil {
			c.Observability.Sampling.Ratio = fv
		}
	}

	// Plugins & Memory
	if v := os.Getenv("DETECTVIZ__PLUGIN__PATHS"); v != "" {
		c.Plugin.Paths = splitCSV(v)
	}
	setStr(&c.Plugin.Registry, os.Getenv("DETECTVIZ__PLUGIN__REGISTRY"))

	setStr(&c.Memory.Backend, os.Getenv("DETECTVIZ__MEMORY__BACKEND"))
	setStr(&c.Memory.DSN, os.Getenv("DETECTVIZ__MEMORY__DSN"))
	if v := os.Getenv("DETECTVIZ__MEMORY__DEFAULT_TTL_SECONDS"); v != "" {
		if iv, err := atoi(v); err == nil {
			c.Memory.DefaultTTLSeconds = iv
		}
	}
}

func setStr(dst *string, v string) {
	if v != "" {
		*dst = v
	}
}

func atoi(s string) (int, error) {
	var x int
	_, err := fmt.Sscanf(s, "%d", &x)
	return x, err
}

func parseBool(s string) (bool, error) {
	s = strings.ToLower(strings.TrimSpace(s))
	switch s {
	case "1", "true", "t", "yes", "y", "on":
		return true, nil
	case "0", "false", "f", "no", "n", "off":
		return false, nil
	default:
		return false, fmt.Errorf("invalid bool: %s", s)
	}
}

func splitCSV(s string) []string {
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

// atof 解析浮點數（用於 sampling.ratio）
func atof(s string) (float64, error) {
	var f float64
	_, err := fmt.Sscanf(strings.TrimSpace(s), "%f", &f)
	return f, err
}

// parseHeadersCSV 解析以逗號分隔的 k=v 列表到 map
// 例如："authorization=Bearer xxx, x-foo=bar" → {"authorization":"Bearer xxx", "x-foo":"bar"}
func parseHeadersCSV(s string) map[string]string {
	out := map[string]string{}
	for _, part := range strings.Split(s, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		kv := strings.SplitN(part, "=", 2)
		if len(kv) != 2 {
			continue
		}
		k := strings.TrimSpace(kv[0])
		v := strings.TrimSpace(kv[1])
		if k != "" {
			out[k] = v
		}
	}
	return out
}
